datamapper/changeset/validators.py

```diff
@@ -0,0 +1,99 @@
+from typing import Any, Callable, Dict, List, Optional, Tuple, Type, Union
+
+from datamapper.changeset.types import FieldValidator
+
+TypeValidator = Callable[[Any], bool]
+Errors = Dict[str, List[str]]
+
+
+def human_type_name(type_name: str) -> str:
+    return {"int": "integer", "str": "string"}.get(type_name, type_name)
+
+
+def _add_error(errors: Errors, key: str, msg: str) -> Errors:
+    errors_ = dict(errors)
+    existing_errors = errors.get(key, [])
+    errors_[key] = [*existing_errors, msg]
+    return errors_
+
+
+def validate_types(params: dict, types: dict, data: dict) -> Tuple[dict, dict]:
+    errors: dict = {}
+    changes = dict(params)
+
+    for (key, val) in params.items():
+        schema_type = types.get(key)
+        if schema_type is None:
+            changes.pop(key)
+            continue
+
+        error = _validate_type(schema_type, val)
+        if error is not None:
+            errors = _add_error(errors, key, error)
+            changes.pop(key)
+
+    return (changes, errors)
+
+
+def _validate_type(schema_type: Type, val: Any) -> Optional[str]:
+    if val is None:
+        return None
+
+    (error_msg, type_validator) = _get_type_validator(schema_type)
+    return error_msg if not type_validator(val) else None
+
+
+def _get_type_validator(schema_type: Type) -> Tuple[str, TypeValidator]:
+    python_type = _to_python_type(schema_type)
+    error_msg = f"Not a valid {human_type_name(python_type.__name__)}."
+
+    def _validator(v: Any) -> bool:
+        return isinstance(v, python_type)
+
+    return (error_msg, _validator)
+
+
+def _to_python_type(type_: Any) -> Type:
+    try:
+        # For sqlalchemy types. Converts to its corresponding python type.
+        return type_.python_type
```

Comment:
Should this bit happen in the `ModelDataChangesetWrapper`? If that were the case, we could make it so that `validate_types` would only ever receive python types.

Reply:
Good idea!